tsdb/store.go

```diff
@@ -1496,7 +1496,21 @@ func (s *Store) DeleteSeries(database string, sources []influxql.Source, conditi
 		// No series file means nothing has been written to this DB and thus nothing to delete.
 		return nil
 	}
-	shards := s.filterShards(byDatabase(database))
+
+	shardFilter := byDatabase(database)
+	if len(sources) != 0 {
+		if measurement, ok := sources[0].(*influxql.Measurement); ok {
+			if measurement.RetentionPolicy != "" {
+				shardFilter = func(sh *Shard) bool {
```

Comment:
I think it would be cleaner here to actually make a composed function that calls the first shardFilter function. That way we can change `byDatabase` to do more or less or different things.

Reply:
Something like this:

```
func ComposeFilter(original func(i int) bool, addition func(i int) bool) func(i int) bool {

	return func(i int) bool { return original(i) && addition(i) }
}
```

Reply:
I do like that suggestion, I think it looks really clean.

Reply:
`ComposeFilter` seems needlessly complicated, especially since it would get used one place.

You could just make use `byDatabase` in the new `shardFilter` definition:
```go
shardFilter := func(sh *Shard) bool { return byDatabase(sh) && sh.retentionPolicy == measurement.RetentionPolicy }
```

You could then pull `sh.retentionPolicy == measurement.RetentionPolicy` into a `byRetentionPolicy` function if you wanted.

Reply:
I would probably change the name of `shardFilter` to something else because it is too similar to `filterShards`. My brain crossed them together the first time I read the code. Maybe change to `shardFilter` to `sf` or `filter`.

Reply:
Question: What if `len(sources) >= 2`? Is this illegal and something we should return an error on? Or should be handle this properly?

My intuition is that this is illegal and should return an error.

Reply:
Do we allow wildcards in DELETEs?  That might produce lots of sources. I don't remember.

Reply:
Let me check that out and see if I can make it so multiple sources get passed --- afaik in the influxQL pr I pass some regex operators `*` specifically for wildcard handling and that fails for the parser.

Reply:
Okay so with a wildcard the query parser fails:
```
> delete from test*."average_temperature" where time > '2019-08-16T00:06:00Z'
ERR: error parsing query: found *, expected ; at line 1, char 17
```

And if you wrap it in quotes it just treats it like a string so "test*" == test*

Reply:
Also I removed the compose filter fn.

@gwossum unfortunately when attempting to do something like: 

```
				shardFilterFn = func(sh *Shard) bool {
					return byDatabase(database) && byRetentionPolicy(measurement.RetentionPolicy)
				}
```

I'm getting and error that `func(sh *Shard) bool does not implement the && operator` 

I think I will leave it as is so that its relatively simple to read.

Reply:
We do support wildcards for measurements in DELETE.  Please verify the syntax to use, and add tests including them.

Reply:
> We do support wildcards for measurements in DELETE. Please verify the syntax to use, and add tests including them.

Sounds good 🫡

Reply:
@devanbenz You could do `byDatabase(database)(sh) && byRetentionPolicy(retention)(sh)`, but it's probably better to call `byDatabase()` and `byRetentionPolicy()` outside the lambda and assign them to variables, then use them inside the lambda. Or use @davidby-influx's `ComposeFilter` idea. Another way to do the compose filter would be something like:

```go
func ComposeFilters(fns ...func(sh *Shard) bool) func(sh *Shard) bool {
   return func(sh *Shard) bool {
        for _, f := range fns {
            if !f(sh) {
                return false
            }
        }
        return true
    }
}

....

    shards := s.filterShards(ComposeFilters(byDatabase(db), byRetentionPolicy(rp)))
````

I find that way more readable, but that's personal preference.

I'd also be tempted to add a `type ShardPredicate func(sh *Shard) bool` :grin:

Reply:
Ah yes! @gwossum this is similar to what I was *attempting* to go for with my earlier design of passing in multiple functions but much much more elegant 🙏

Reply:
Okay modifications have been made -- testing of wildcards as well as adding the composable function for shard filtering + I refactored other areas that is calling the function to include the composition fn